lib/datadog/core/utils/only_once_successful.rb

```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+require_relative 'only_once'
+
+module Datadog
+  module Core
+    module Utils
+      # Helper class to execute something with only one success.
+      #
+      # This is useful for cases where we want to ensure that a block of code is only executed once, and only if it
+      # succeeds. One such example is sending app-started telemetry event.
+      #
+      # Successful execution is determined by the return value of the block: any truthy value is considered success.
+      #
+      # Thread-safe when used correctly (e.g. be careful of races when lazily initializing instances of this class).
+      #
+      # Note: In its current state, this class is not Ractor-safe.
+      # In https://github.com/DataDog/dd-trace-rb/pull/1398#issuecomment-797378810 we have a discussion of alternatives,
+      # including an alternative implementation that is Ractor-safe once spent.
+      class OnlyOnceSuccessful < OnlyOnce
```

Comment:
I really like this feature. It's a nice little improvement.

Thinking more expansively, is it worth adding retry/limit functionality to this? (e.g. allow it to try X times before marking as "ran") I'm thinking there may be some scenarios where we don't want to try something expensive/failure prone indefinitely.

If this is not of benefit to this feature, then it doesn't need to be implemented now. Maybe its something we could implement later.

Reply:
yes, I think this is a nice feature, I will add limit to this class

Reply:
The only issue with it would be this one: we don't want this class to be in "successful" state afterwards. In telemetry case it will start then sending heartbeat events even though app-started failed. I will introduce a new "failed" state to this class.